src/basic-noise-qvm.lisp

```diff
@@ -0,0 +1,194 @@
+         (kraus-ops (list t1-ops t2-ops depol-ops)))
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple qubits) params)
+    (apply-all-kraus-maps qvm instr kraus-ops)
+    (incf (pc qvm))
+    qvm))
+
+
+(defmethod apply-classical-readout-noise ((qvm basic-noise-qvm) (instr quil:measure))
+  ;; Apply the classical readout noise to the state of an BASIC-NOISE-QVM
+  (%corrupt-qvm-memory-with-povm qvm instr (readout-povms qvm)))
+
+
+(defun tphi (t1 t2)
+  "Calculate t_phi from T1 and T2. t_phi = (2*t1*t2) / (2*t1 + t2). If both T1 and T2 are 0, TPHI = 0."
+  (if (and (plusp t1) (plusp t2))
+      (/ (* 2 (* t1 t2))
+         (+ t2 (* 2 t1)))
+      0))
+
+
+(defun generate-damping-kraus-map (t1 gate-time)
+  "Given a value for T1 and a GATE-TIME, generates the kraus operators corresponding to the t1 noise."
+  (let* ((prob (/ gate-time t1))
+         (k0 (magicl:make-complex-matrix 2 2 (list 0 0 (sqrt prob) 0)))
+         (k1 (magicl:make-complex-matrix 2 2 (list 1 0 0  (sqrt (- 1 prob))))))
+    (list k0 k1)))
+
+
+(defun generate-dephasing-kraus-map (t-phi gate-time)
+  "Given a value for T-PHI (dephasing time) and a GATE-TIME, calculates the kraus operators corresponding to the dephasing noise."
+  (let*  ((prob (/ gate-time t-phi))
+          (prob-k0 (/ prob 2))
+          (prob-k1 (- 1 prob-k0))
+          (base-matrices '("I" "Z")))
+    (generate-kraus-ops base-matrices (list (sqrt prob-k0) (sqrt prob-k1)))))
+
+
+(defun generate-depolarizing-kraus-map (prob)
+  "Given a probability of depolarization PROB, calculate the kraus operation elements for the depolarizing channel."
+  (assert (< 0 prob 1) (prob) "DEPOL-PROB must be between 0 and 1")
+  (let* ((pk0 (sqrt (- 1 (/ (* 3 prob) 4))))
+         (pkn (sqrt (/ prob 4)))
+         (base-matrices '("I" "X" "Y" "Z"))
+         (probs (list pk0 pkn pkn pkn)))
+    (generate-kraus-ops base-matrices probs)))
+
+
+;; TODO: Need to double check how damping dephasing kraus ops are
+;; calculated before just combining it into
+;; generate-dephasing-kraus-ops. For now I am pretty sure the
+;; calculations are the same.
+(defun generate-damping-dephasing-kraus-map (t2 gate-time)
+  "Calculates the kraus operators for dephasing and damping noise from T2, which is decoherence time."
+  (generate-dephasing-kraus-map t2 gate-time))
+
+
+(defun kraus-kron (k1s k2s)
+  "Calculate the tensor product of two kraus maps K1 and K2 by tensoring their elems. If one of the kraus maps is NIL, tensor the other with the identity matrix."
+  (cond ((not k1s) (loop :for k :in k2s
+                         :collect (magicl:kron (magicl:make-identity-matrix 2) k )))
+        ((not k2s) (loop :for k :in k1s
+                         :collect (magicl:kron k (magicl:make-identity-matrix 2))))
+        (t  (loop :for k1 :in k1s :append (loop :for k2 :in k2s
+                                                :collect (magicl:kron k1 k2))))))
```

Comment:
Use NULL instead of NOT and compute the 2x2 identity just once:
```
  (cond ((null k1s) (loop :with identity-matrix := (magicl:make-identity-matrix 2)
                          :for k :in k2s :collect (magicl:kron identity-matrix k)))
        ((null k2s) (loop :with identity-matrix := (magicl:make-identity-matrix 2)
                          :for k :in k1s :collect (magicl:kron k identity-matrix)))
        (t (loop :for k1 :in k1s :append (loop :for k2 :in k2s
                                               :collect (magicl:kron k1 k2))))))
```